ballerina/graphql_client.bal

```diff
@@ -0,0 +1,183 @@
+    remote isolated function execute(string query, map<anydata>? variables = (), map<string|string[]>? headers = (), 
+                                     typedesc<GraphQLGenericResponse|GraphQLDataResponse|record{}|json> targetType = <>) 
+                                     returns targetType|GError = @java:Method {
+        'class: "io.ballerina.stdlib.graphql.client.QueryExecutor",
+        name: "execute"
+    } external;
+
+    private isolated function processExecute(typedesc<GraphQLGenericResponse|GraphQLDataResponse|record{}|json> targetType, 
+                                             string query, map<anydata>? variables, map<string|string[]>? headers) 
+                                             returns GraphQLGenericResponse|GraphQLDataResponse|record{}|json|GError {
+        http:Request request = new;
+        json graphqlPayload = getGraphqlPayload(query, variables);
+        request.setPayload(graphqlPayload);
+
+        json|http:ClientError httpResponse = self.clientEp->post("", request, headers = headers);
+
+        do {
+            if httpResponse is http:ClientError {
+                if (httpResponse is http:ApplicationResponseError) {
+                    anydata data = check httpResponse.detail().get("body").ensureType(anydata);
+                    return error ClientError("GraphQL Client Error", body = data);
+                }
+                return error ClientError("GraphQL Client Error", httpResponse);
+            } else {
+                map<json> responseMap = <map<json>> httpResponse;
+
+                if targetType is typedesc<GraphQLGenericResponse> {
+                    GraphQLGenericResponse response = check responseMap.cloneWithType(targetType);
+                    return response;
+                } else if targetType is typedesc<GraphQLDataResponse> {
+                    json responseData = responseMap.get("data");
+                    if (responseMap.hasKey("extensions")) {
+                        responseData = check responseData.mergeJson({ "__extensions" : responseMap.get("extensions") });
+                    }
+                    if (responseMap.hasKey("errors")) {
+                        responseData = check responseData.mergeJson({ "__errors" : responseMap.get("errors") });
+                    }
+                    GraphQLDataResponse response = check responseData.cloneWithType(targetType);
+                    return response;
+                } else if targetType is typedesc<record{}> {
+                    record{} response = check httpResponse.cloneWithType(targetType);
+                    return response;
+                } else if targetType is typedesc<json> {
+                    json response = check httpResponse.cloneWithType(targetType);
+                    return response;
+                } else {
+                    return error ClientError("GraphQL Client Error, Invalid binding type.");
+                }
+            }
+        } on fail var e {
```

Comment:
Ditto.

Reply:
We use the `do-on fail` block to catch all the generic response processing errors and to return them as a `ClientError` which we renamed as a `RequestError`.